### PR TITLE
fix(react-positioning): reset overrideTargetRef when target from hook option is undefined

### PR DESCRIPTION
**Pull request by YuanboXue-Amber**
### Current/Expected Behavior:
Example - a menu with both `openOnContext` and `openOnHover` prop: https://codesandbox.io/s/muddy-pine-i9tnb7?file=/example.tsx
Repro step:
1. right click on trigger to open context menu
2. click document to close context menu
3. hover on trigger to open menu again, notice that the menu popover is opened relative to where mouse click happened in step 1. The expected behavior is to have menu opened relative to the trigger.

### Fix:
In `useMenu`, when context menu opens, we set `contextTarget` to the position of mouse click; when context menu closes, we [reset `contextTarget` to `undefined`](https://github.com/microsoft/fluentui/blob/d59683655d4d2a3775df4a2b41a09504cddd72ad/packages/react-components/react-menu/src/components/Menu/useMenu.tsx#L184). 
`contextTarget` is passed to `usePositioning` hook. But in `usePositioning` hook, we ignore the case when `contextTarget` is reset to undefined

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports                                                                                                                                                             |      Baseline (minified/GZIP) |                            PR |                                                                                                                                                                                                Change |
| :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: | ----------------------------: | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Combobox.fixture.js'>Combobox (including child components)</abbr>                                                 |  `75.071 kB`<br />`24.394 kB` |  `75.083 kB`<br />`24.402 kB` | `12 B` <br />`8 B`  |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Dropdown.fixture.js'>Dropdown (including child components)</abbr>                                                 |  `74.799 kB`<br />`24.352 kB` |  `74.811 kB`<br />`24.358 kB` | `12 B` <br />`6 B`  |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `189.227 kB`<br />`52.474 kB` | `189.239 kB`<br />`52.482 kB` | `12 B` <br />`8 B`  |
| <samp>react-menu</samp> <br /> <abbr title='bundle-size/Menu.fixture.js'>Menu (including children components)</abbr>                                                          | `116.558 kB`<br />`35.725 kB` |  `116.57 kB`<br />`35.729 kB` | `12 B` <br />`4 B`  |
| <samp>react-menu</samp> <br /> <abbr title='bundle-size/Menu.Selectable.fixture.js'>Menu (including selectable components)</abbr>                                             | `119.627 kB`<br />`36.246 kB` |  `119.639 kB`<br />`36.25 kB` | `12 B` <br />`4 B`  |
| <samp>react-popover</samp> <br /> <abbr title='bundle-size/Popover.fixture.js'>Popover</abbr>                                                                                 |   `102.756 kB`<br />`31.5 kB` | `102.768 kB`<br />`31.505 kB` | `12 B` <br />`5 B`  |
| <samp>react-positioning</samp> <br /> <abbr title='bundle-size/usePopper.fixture.js'>usePositioning</abbr>                                                                    |     `19.7 kB`<br />`7.404 kB` |   `19.712 kB`<br />`7.408 kB` | `12 B` <br />`4 B`  |
| <samp>react-tooltip</samp> <br /> <abbr title='bundle-size/Tooltip.fixture.js'>Tooltip</abbr>                                                                                 |  `41.566 kB`<br />`14.644 kB` |  `41.578 kB`<br />`14.649 kB` | `12 B` <br />`5 B`  |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                |         Size (minified/GZIP) |
| ---------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------: |
| <samp>react-alert</samp> <br /> <abbr title='bundle-size/Alert.fixture.js'>Alert</abbr>                                                                          | `84.529 kB`<br />`21.171 kB` |
| <samp>react-avatar</samp> <br /> <abbr title='bundle-size/Avatar.fixture.js'>Avatar</abbr>                                                                       | `48.828 kB`<br />`13.842 kB` |
| <samp>react-avatar</samp> <br /> <abbr title='bundle-size/AvatarGroup.fixture.js'>AvatarGroup</abbr>                                                             |   `14.95 kB`<br />`5.989 kB` |
| <samp>react-avatar</samp> <br /> <abbr title='bundle-size/AvatarGroupItem.fixture.js'>AvatarGroupItem</abbr>                                                     | `68.796 kB`<br />`19.182 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr> | `62.781 kB`<br />`17.574 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>               |   `33.4 kB`<br />`11.008 kB` |
| <samp>react-persona</samp> <br /> <abbr title='bundle-size/Persona.fixture.js'>Persona</abbr>                                                                    | `53.946 kB`<br />`15.225 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr>                                            |   `5.857 kB`<br />`1.978 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/d59683655d4d2a3775df4a2b41a09504cddd72ad'>d59683655d4d2a3775df4a2b41a09504cddd72ad</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/Avatar-mount_base.html">1311</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/Avatar-mount.html">1279</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/Button-mount_base.html">933</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/Button-mount.html">942</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/FluentProvider-mount_base.html">1569</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/FluentProvider-mount.html">1559</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/FluentProviderWithTheme-mount_base.html">579</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/FluentProviderWithTheme-mount.html">578</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">541</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">546</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">572</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">583</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/MakeStyles-mount_base.html">1960</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/MakeStyles-mount.html">1951</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/SpinButton-mount_base.html">2420</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25135/perf-test-react-components/SpinButton-mount.html">2309</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Review by bsunderhus — APPROVED** [review]

File: apps/vr-tests-react-components/src/stories/Positioning.stories.tsx [hunk 1]
```diff
@@ -442,6 +442,29 @@ const VirtualElement = () => {
...
+
+  React.useEffect(() => {
+    if (virtualElement) {
+      setVirtualElement(null);
+    }
+  }, [virtualElement]);
+
+  return (
+    <>
+      <button ref={setVirtualElement}>Virtual Target</button>
```
Comment (ling1726) [inline apps/vr-tests-react-components/src/stories/Positioning.stories.tsx:461]: nit: maybe have the virtual target on top of the target, that way it should be easier to see that the screenshot is incorrect beacause the popover on the virtual target will be covering the target

File: packages/react-components/react-positioning/src/usePositioning.ts [hunk 2]
```diff
@@ -125,6 +125,8 @@ export function usePositioning(
   useIsomorphicLayoutEffect(() => {
     if (options.target) {
       overrideTargetRef.current = options.target;
+    } else {
+      overrideTargetRef.current = null;
     }
```
Comment (ling1726) [inline packages/react-components/react-positioning/src/usePositioning.ts:130]: Isn't this the same as the below one liner ?
```ts
overrideTargetRef.current = options.target;
```

**Comment by ling1726** [comment]
The `usePopperMouseTarget` was implemented before our imperative API was a thing. I think it's actually a more scalable fix to refactor the `Menu` and `Popover` components to use `positioningRef.current.setTarget` in this case, it would actually avoid saving something to state since `usePopperMouseTarget` saves the target in state

File: packages/react-components/react-positioning/src/usePositioning.ts (hunk 2 above)
Comment (YuanboXue-Amber) [reply packages/react-components/react-positioning/src/usePositioning.ts]: updated :)

**Comment by YuanboXue-Amber** [comment]
> The `usePopperMouseTarget` was implemented before our imperative API was a thing. I think it's actually a more scalable fix to refactor the `Menu` and `Popover` components to use `positioningRef.current.setTarget` in this case, it would actually avoid saving something to state since `usePopperMouseTarget` saves the target in state

had a 1to1 chat about it. This would be good to have, I'll tackle it in a separate PR

**Pull request merged**